Readme.md
# Sitemap (ext:nkwsitemap)

## Funktion

* Stellt eine Sitemap auf einer Seite bereit.
* Zeigt den Seitenbaum ab einer freiwählbaren Seite an
* Fällt bei fehlernder Übersetzung einer Seite auf Language 0 zurück

## Konfiguration (Anwendung)

* Einfügen des Plugins auf einer beliebigen Seite
* Angabe des Startseite verpflichtend


ChangeLog
2011-07-06 Ingo Pfennigstorf <pfennigstorf@sub.uni-goettingen.de>
	* [Bugfix] Anzeige der Sitemap Aufrufe der nkwlib (WWW-685)

2011-06-20 Ingo Pfennigstorf <pfennigstorf@sub.uni-goettingen.de>
	* Statische Aufrufe von nkwlib implementiert

2010-09-09 - 0.0.9 - Nils K. Windisch <windisch@sub.uni-goettingen.de>
	* bug fix in ext_emconf.php
	* added file ChangeLog

ext_tables.php
<?php
/***************************************************************
 *  Copyright notice
 *
 *  (c) 2010 Nils K. Windisch <windisch@sub.uni-goettingen.de>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/
if (!defined('TYPO3_MODE')) {
	die ('Access denied.');
}
$TCA['tt_content']['types']['list']['subtypes_excludelist'][$_EXTKEY . '_pi1'] = 'layout,select_key';
// add plugin
\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addPlugin(
		array(
				'LLL:EXT:nkwsitemap/locallang_db.xml:tt_content.list_type_pi1',
				$_EXTKEY . '_pi1',
				\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::extRelPath($_EXTKEY) . 'ext_icon.gif'
		),
		'list_type'
);


ext_localconf.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2010 Nils K. Windisch <windisch@sub.uni-goettingen.de>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/
if (!defined ('TYPO3_MODE')) {
 	die ('Access denied.');
}

// add plugin
\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addPItoST43($_EXTKEY, 'pi1/class.tx_nkwsitemap_pi1.php', '_pi1', 'list_type', 1);
?>

pi1/class.tx_nkwsitemap_pi1.php
<?php

/* * *************************************************************
 *  Copyright notice
 *
 *  (c) 2009 Nils K. Windisch <windisch@sub.uni-goettingen.de>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 * ************************************************************* */

/**
 * Plugin 'Custom Sitemap' for the 'nkwsitemap' extension.
 */
class tx_nkwsitemap_pi1 extends \TYPO3\CMS\Frontend\Plugin\AbstractPlugin {

	public $prefixId = 'tx_nkwsitemap_pi1';
	public $scriptRelPath = 'pi1/class.tx_nkwsitemap_pi1.php';
	public $extKey = 'nkwsitemap';
	public $pi_checkCHash = TRUE;

	/**
	 * @var \TYPO3\CMS\Frontend\Page\PageRepository
	 */
	protected $pageRepository;

	/**
	 * @var \TYPO3\CMS\Core\Database\DatabaseConnection
	 */
	protected $db;

	/**
	 * The main method of the PlugIn
	 *
	 * @param string $content The PlugIn content
	 * @param array $conf The PlugIn configuration
	 * @return string The content that is displayed on the website
	 */
	public function main($content, $conf) {
		$this->conf = $conf;

		$this->pageRepository = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance(\TYPO3\CMS\Frontend\Page\PageRepository::class);
		$this->db = $GLOBALS['TYPO3_DB'];

		$this->pi_setPiVarDefaults();
		$content .= $this->displayTree($this->getPageTreeIds($this->cObj->data['pages']));
		return $this->pi_wrapInBaseClass($content);
	}

	/**
	 * Get the IDds of a pagetree as Array
	 *
	 * @param int $startId
	 * @return array
	 */
	protected function getPageTreeIds($startId) {

		$tree = array();

		$res1 = $this->db->exec_SELECTquery(
				'uid',
				'pages',
				'pid = ' . $startId . ' AND deleted = 0 AND hidden = 0 AND pid > 0 AND t3ver_wsid = 0',
				'',
				'sorting ASC',
				'');
		while ($row1 = $this->db->sql_fetch_assoc($res1)) {
			$children = $this->getPageTreeIds($row1['uid']);
			if ($children) {
				$tree[$row1['uid']]['children'] = $this->getPageTreeIds($row1['uid']);
			} else {
				$tree[$row1['uid']]['children'] = 0;
			}
		}
		return $tree;
	}

	/**
	 * Displays the tree for the sitemap
	 *
	 * @param array $tree
	 * @return string
	 */
	protected function displayTree($tree) {
		$displayTree = '<ul>';
		foreach ($tree as $uid => $value) {

			$page = $this->pageRepository->getPage($uid);

			if ($GLOBALS['TSFE']->sys_language_uid !== 0) {
				$page = $this->pageRepository->getPageOverlay($page, $GLOBALS['TSFE']->sys_language_uid);
			}

			$title = $page['title'];
			$displayTree .= '<li>';
			// url title hacks
			$saveATagParams = $GLOBALS['TSFE']->ATagParams;
			$GLOBALS['TSFE']->ATagParams = 'title="' . $title . '"';
			$displayTree .= $this->pi_LinkToPage($title, $uid, '', '');
			$GLOBALS['TSFE']->ATagParams = $saveATagParams;

			if ($value['children']) {
				$displayTree .= $this->displayTree($value['children']);
			}
			$displayTree .= '</li>';
		}
		$displayTree .= '</ul>';
		return $displayTree;
	}

}

if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/' . $extKey . '/pi1/class.tx_nkwsitemap_pi1.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/' . $extKey . '/pi1/class.tx_nkwsitemap_pi1.php']);
}


Classes/Controller/SitemapController.php
<?php
namespace Subugoe\NkwSitemap\Controller;

/* * *************************************************************
 *  Copyright notice
 *
 *  (c) 2009 Nils K. Windisch <windisch@sub.uni-goettingen.de>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 * ************************************************************* */
require_once(\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::extPath('nkwlib') . 'class.tx_nkwlib.php');

/**
 * Plugin 'Custom Sitemap' for the 'nkwsitemap' extension.
 */
class SitemapController extends \TYPO3\CMS\Frontend\Plugin\AbstractPlugin {

	public $prefixId = 'SitemapController';
	public $scriptRelPath = 'Classes/Controller/SitemapController.php';
	public $extKey = 'nkwsitemap';
	public $pi_checkCHash = TRUE;

	/**
	 * The main method of the PlugIn
	 *
	 * @param string $content The PlugIn content
	 * @param array $conf The PlugIn configuration
	 * @return string The content that is displayed on the website
	 */
	public function main($content, $conf) {
		$this->conf = $conf;
		$this->pi_setPiVarDefaults();
		$content = $this->displayTree(\tx_nkwlib::getPageTreeIds($this->cObj->data['pages']), $GLOBALS['TSFE']->sys_page->sys_language_uid);
		return $this->pi_wrapInBaseClass($content);
	}

	/**
	 * Displays the tree for the sitemap
	 *
	 * @param string $tree
	 * @param int $lang
	 * @return string
	 */
	protected function displayTree($tree, $lang) {
		$displayTree = '<ul>';
		foreach ($tree as $key => $value) {
			$title = \tx_nkwlib::getPageTitle($key, $lang);
			$displayTree .= '<li>';
			// url title hacks
			$saveATagParams = $GLOBALS['TSFE']->ATagParams;
			$GLOBALS['TSFE']->ATagParams = 'title="' . $title . '"';
			$displayTree .= $this->pi_LinkToPage($title, $key, '', '');
			$GLOBALS['TSFE']->ATagParams = $saveATagParams;

			if ($value['children']) {
				$displayTree .= $this->displayTree($value['children'], $lang);
			}
			$displayTree .= '</li>';
		}
		$displayTree .= '</ul>';
		return $displayTree;
	}
}


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "nkwsitemap".
#
# Auto generated 06-07-2011 09:10
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Custom Sitemap',
	'description' => '',
	'category' => 'plugin',
	'author' => 'Nils K. Windisch',
	'author_email' => 'windisch@sub.uni-goettingen.de',
	'shy' => '',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'beta',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => 'Goettingen State and University Library, Germany http://www.sub.uni-goettingen.de',
	'version' => '2.0.0',
	'constraints' => array(
		'depends' => array(
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:9:{s:9:"ChangeLog";s:4:"5e5e";s:10:"README.txt";s:4:"45ec";s:12:"ext_icon.gif";s:4:"c8ca";s:17:"ext_localconf.php";s:4:"2ed5";s:14:"ext_tables.php";s:4:"87f4";s:16:"locallang_db.xml";s:4:"8ebe";s:19:"doc/wizard_form.dat";s:4:"d6d8";s:20:"doc/wizard_form.html";s:4:"bc49";s:31:"pi1/class.tx_nkwsitemap_pi1.php";s:4:"be66";}',
	'suggests' => array(
	),
);

?>